### PR TITLE
src/hmem_ze, prov/shm: dynamically allocate device resource, create mapping between physical devices and user space devices

### PR DESCRIPTION
**Pull request by aingerson**
-dynamically allocate device info and resource to remove hard coded maximum
-create mapping between physical devices and user space devices to fix shm IPC fd mapping
-fix copy engine environment variable

### CODE REVIEW

**Comment by shijin-aws** [comment]
Not related to the PR, but I am thinking whether we can make shm max peers configurable and smr_region dynamically allocated.. Currently this happens in ep enable so I think we should be able to make it configured via fi_setopt..

File: src/hmem_ze.c [hunk 1]
```diff
@@ -48,6 +48,7 @@
 
 struct ofi_ze_dev_info {
 	ze_command_queue_handle_t cmd_queue;
+	int device;
```
Comment (j-xiong) [inline src/hmem_ze.c:51]: Could use a different name to avoid confusion. For example, pci_device.

File: src/hmem_ze.c [hunk 2]
```diff
@@ -443,11 +446,12 @@ int ze_hmem_get_shared_handle(int dev_fd, void *dev_buf, int *ze_fd,
...
 }
 
-int ze_hmem_open_shared_handle(int dev_fd, void **handle, int *ze_fd,
-			       uint64_t device, void **ipc_ptr)
+int ze_hmem_open_shared_handle(uint64_t device, int *peer_fds, void **handle,
+			       int *ze_fd, void **ipc_ptr)
 {
 	struct drm_prime_handle open_fd = {0, 0, 0};
 	ze_ipc_mem_handle_t ze_handle;
+	int dev_fd = peer_fds[dev_info[device].device];
```
Comment (j-xiong) [inline src/hmem_ze.c:454]: Here is an example of how the two `device` terms can get confusing.

File: src/hmem_ze.c [hunk 3]
```diff
@@ -849,6 +854,10 @@ int ze_hmem_init(void)
 
 		dev_info[num_devices].uuid = dev_prop.uuid;
 
+		if (!dev_prop.subdeviceId)
+			pci_dev++;
+		dev_info[num_devices].device = pci_dev;
```
Comment (j-xiong) [inline src/hmem_ze.c:859]: This may not work as expected when ZE_AFFINITY_MASK is set.

Comment (j-xiong) [reply src/hmem_ze.c:859]: can find the correct pci device by comparing uuid.

Comment (aingerson) [reply src/hmem_ze.c:859]: I'm not really sure how to find it with the uuid. That's a 16 byte uint8 array that seems to be unique for each tile. For example on my system:
```
{id = "\206\200\326\v/\000\000\000:\000\000\000\000\000\000\001"}
{id = "\206\200\326\v/\000\000\000:\000\000\000\000\000\000\002"}
{id = "\206\200\326\v/\000\000\000\232\000\000\000\000\000\000\001"}
{id = "\206\200\326\v/\000\000\000\232\000\000\000\000\000\000\002"}
```

**Comment by aingerson** [comment]
@shijin-aws Totally! It's on my list of things to do and I'm planning on working it into my new shm changes :)

File: src/hmem_ze.c (hunk 3 above)
Comment (j-xiong) [reply src/hmem_ze.c:859]: The bytes around the 9th byte are the pci address of the device. Not sure where is the start and end.
Note that ":" is 0x3a and "\232" is 0x9a. The pci addresses are `0000:3a:00.0` and `0000:9a:00.0`.

Comment (aingerson) [reply src/hmem_ze.c:859]: Right I see that. But how do I map ":" and "\232" to 0 and 1? Is there a way to lookup all ZE PCI devices (regardless of the affinity mask)?

Comment (aingerson) [reply src/hmem_ze.c]: @j-xiong See update which matches it based on the PCI bus ID. This isn't the prettiest thing ever but it does the job...

File: src/hmem_ze.c [hunk 4]
```diff
@@ -403,26 +407,36 @@ static int ze_hmem_init_fds(void)
...
+		if (dev_fds[num_pci_devices] == -1)
 			goto err;
-		i++;
+		uuid_str = strtok(ent->d_name, ":");
+		uuid_str = strtok(NULL, ":");
+		uuid = (uint8_t) strtol(uuid_str, NULL, 16);
+
+		for (i = 0; i < num_devices; i++) {
+			if (dev_info[i].uuid.id[8] == uuid)
+				dev_info[i].pci_device = num_pci_devices;
```
Comment (j-xiong) [inline src/hmem_ze.c:419]: For the full pci address `domain:bus:dev.func`, the `domain` part needs to be considered as well. For example, here is a list of pci addresses on an actual machine with 6 PVCs: `0000:18:00.0`, `0000:42:00.0`, `0000:6c:00.0`, `0001:18:00.0`, `0001:42:00.0`, `0001:6c:00.0`.

Comment (aingerson) [reply src/hmem_ze.c]: @j-xiong Updated. Do the dev and func need to be considered as well?

Comment (j-xiong) [reply src/hmem_ze.c]: So far dev and func are always 0 and can be ignored.

**Review by j-xiong — APPROVED** [review]

**Pull request merged**